File: 6/enigma.py
import sys

alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

def mod(n, m):
    return ((n % m) + m) % m

ROTORS = [
    {'wiring': 'EKMFLGDQVZNTOWYHXUSPAIBRCJ', 'notch': 'Q'},  # Rotor I
    {'wiring': 'AJDKSIRUXBLHWTMCQGZNPYFVOE', 'notch': 'E'},  # Rotor II
    {'wiring': 'BDFHJLCPRTXVZNYEIWGAKMUSQO', 'notch': 'V'},  # Rotor III
]
REFLECTOR = 'YRUHQSLDPXNGOKMIEBFZCWVJAT'

def plugboard_swap(c, pairs):
    for a, b in pairs:
        if c == a:
            return b
        if c == b:
            return a
    return c

class Rotor:
    def __init__(self, wiring, notch, ring_setting=0, position=0):
        self.wiring = wiring
        self.notch = notch
        self.ring_setting = ring_setting
        self.position = position

    def step(self):
        self.position = mod(self.position + 1, 26)

    def at_notch(self):
        return alphabet[self.position] == self.notch

    def forward(self, c):
        idx = mod(alphabet.index(c) + self.position - self.ring_setting, 26)
        return self.wiring[idx]

    def backward(self, c):
        idx = self.wiring.index(c)
        return alphabet[mod(idx - self.position + self.ring_setting, 26)]

class Enigma:
    def __init__(self, rotor_ids, rotor_positions, ring_settings, plugboard_pairs):
        self.rotors = [
            Rotor(
                ROTORS[id]['wiring'],
                ROTORS[id]['notch'],
                ring_settings[i],
                rotor_positions[i]
            )
            for i, id in enumerate(rotor_ids)
        ]
        self.plugboard_pairs = plugboard_pairs

    def step_rotors(self):
        if self.rotors[2].at_notch():
            self.rotors[1].step()
        if self.rotors[1].at_notch():
            self.rotors[0].step()
        self.rotors[2].step()

    def encrypt_char(self, c):
        if c not in alphabet:
            return c
        
        self.step_rotors()
        c = plugboard_swap(c, self.plugboard_pairs)
        
        for rotor in reversed(self.rotors):
            c = rotor.forward(c)
        
        idx = alphabet.index(c)
        c = REFLECTOR[idx]
        
        for rotor in self.rotors:
            c = rotor.backward(c)
        
        return c

    def process(self, text):
        return ''.join(self.encrypt_char(c) for c in text.upper())

def prompt_enigma():
    message = input('Enter message: ')
    pos_str = input('Rotor positions (e.g. 0 0 0): ')
    rotor_positions = [int(x) for x in pos_str.split()]
    
    ring_str = input('Ring settings (e.g. 0 0 0): ')
    ring_settings = [int(x) for x in ring_str.split()]
    
    plug_str = input('Plugboard pairs (e.g. AB CD): ')
    plug_pairs = []
    if plug_str:
        pairs = plug_str.upper().split()
        plug_pairs = [[pair[0], pair[1]] for pair in pairs]

    enigma = Enigma([0, 1, 2], rotor_positions, ring_settings, plug_pairs)
    result = enigma.process(message)
    print('Output:', result)

if __name__ == '__main__':
    if len(sys.argv) > 1:
        # Use default settings for quick testing
        message = sys.argv[1]
        rotor_positions = [0, 0, 0]
        ring_settings = [0, 0, 0]
        plug_pairs = []
        
        enigma = Enigma([0, 1, 2], rotor_positions, ring_settings, plug_pairs)
        result = enigma.process(message)
        print('Input:', message)
        print('Output:', result)
    else:
        prompt_enigma() 

File: 6/enigma.js
const readline = require('readline');

const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
function mod(n, m) {
  return ((n % m) + m) % m;
}

const ROTORS = [
  { wiring: 'EKMFLGDQVZNTOWYHXUSPAIBRCJ', notch: 'Q' }, // Rotor I
  { wiring: 'AJDKSIRUXBLHWTMCQGZNPYFVOE', notch: 'E' }, // Rotor II
  { wiring: 'BDFHJLCPRTXVZNYEIWGAKMUSQO', notch: 'V' }, // Rotor III
];
const REFLECTOR = 'YRUHQSLDPXNGOKMIEBFZCWVJAT';

function plugboardSwap(c, pairs) {
  for (const [a, b] of pairs) {
    if (c === a) return b;
    if (c === b) return a;
  }
  return c;
}

class Rotor {
  constructor(wiring, notch, ringSetting = 0, position = 0) {
    this.wiring = wiring;
    this.notch = notch;
    this.ringSetting = ringSetting;
    this.position = position;
  }
  step() {
    this.position = mod(this.position + 1, 26);
  }
  atNotch() {
    return alphabet[this.position] === this.notch;
  }
  forward(c) {
    const idx = mod(alphabet.indexOf(c) + this.position - this.ringSetting, 26);
    return this.wiring[idx];
  }
  backward(c) {
    const idx = this.wiring.indexOf(c);
    return alphabet[mod(idx - this.position + this.ringSetting, 26)];
  }
}

class Enigma {
  constructor(rotorIDs, rotorPositions, ringSettings, plugboardPairs) {
    this.rotors = rotorIDs.map(
      (id, i) =>
        new Rotor(
          ROTORS[id].wiring,
          ROTORS[id].notch,
          ringSettings[i],
          rotorPositions[i],
        ),
    );
    this.plugboardPairs = plugboardPairs;
  }
  stepRotors() {
    if (this.rotors[1].atNotch()) {
      this.rotors[0].step();
      this.rotors[1].step();
    }
    else if (this.rotors[2].atNotch()) {
      this.rotors[1].step();
    }
    this.rotors[2].step();
  }
  encryptChar(c) {
    if (!alphabet.includes(c)) return c;
    this.stepRotors();
    c = plugboardSwap(c, this.plugboardPairs);
    for (let i = this.rotors.length - 1; i >= 0; i--) {
      c = this.rotors[i].forward(c);
    }

    c = REFLECTOR[alphabet.indexOf(c)];

    for (let i = 0; i < this.rotors.length; i++) {
      c = this.rotors[i].backward(c);
    }

    return c;
  }
  process(text) {
    return text
      .toUpperCase()
      .split('')
      .map((c) => this.encryptChar(c))
      .join('');
  }
}

function promptEnigma() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  rl.question('Enter message: ', (message) => {
    rl.question('Rotor positions (e.g. 0 0 0): ', (posStr) => {
      const rotorPositions = posStr.split(' ').map(Number);
      rl.question('Ring settings (e.g. 0 0 0): ', (ringStr) => {
        const ringSettings = ringStr.split(' ').map(Number);
        rl.question('Plugboard pairs (e.g. AB CD): ', (plugStr) => {
          const plugPairs =
            plugStr
              .toUpperCase()
              .match(/([A-Z]{2})/g)
              ?.map((pair) => [pair[0], pair[1]]) || [];

          const enigma = new Enigma(
            [0, 1, 2],
            rotorPositions,
            ringSettings,
            plugPairs,
          );
          const result = enigma.process(message);
          console.log('Output:', result);
          rl.close();
        });
      });
    });
  });
}

function runTests() {
  console.log('Running Enigma tests...');
  
  const enigma1 = new Enigma([0, 1, 2], [0, 0, 0], [0, 0, 0], []);
  const message1 = 'HELLO';
  const encrypted1 = enigma1.process(message1);
  const enigma2 = new Enigma([0, 1, 2], [0, 0, 0], [0, 0, 0], []);
  const decrypted1 = enigma2.process(encrypted1);
  console.log('Test 1:', message1 === decrypted1 ? 'PASS' : 'FAIL');
  
  const enigma3 = new Enigma([0, 1, 2], [0, 0, 0], [0, 0, 0], [['A', 'B']]);
  const message2 = 'HELLO';
  const encrypted2 = enigma3.process(message2);
  const enigma4 = new Enigma([0, 1, 2], [0, 0, 0], [0, 0, 0], [['A', 'B']]);
  const decrypted2 = enigma4.process(encrypted2);
  console.log('Test 2:', message2 === decrypted2 ? 'PASS' : 'FAIL');
  
  const enigma5 = new Enigma([0, 1, 2], [1, 2, 3], [0, 0, 0], []);
  const message3 = 'HELLO';
  const encrypted3 = enigma5.process(message3);
  const enigma6 = new Enigma([0, 1, 2], [1, 2, 3], [0, 0, 0], []);
  const decrypted3 = enigma6.process(encrypted3);
  console.log('Test 3:', message3 === decrypted3 ? 'PASS' : 'FAIL');
}

if (require.main === module) {
  if (process.argv[2] === '--test') {
    runTests();
  } else {
    promptEnigma();
  }
}
